discourse/discourse
# DEV: convert grant badge modal to component API

#### Context

Part of converting all modals from legacy controller to the new DModal component API. (reference: https://meta.discourse.org/t/using-the-dmodal-api-to-render-modal-windows-aka-popups-dialogs-in-discourse/268304)

This PR updates the grant-badge modal that is used to grant badges for specific posts to the new component API. Updating the component itself to a Glimmer component will be in a separate PR (also calling out a potential footgun for the Glimmer refactor: much of the badge granting logic here is within a Mixin which is no longer supported by glimmer components).

Added a system test to cover this logic end-to-end.

#### Testing:
Existing tests cover the key logic within the GrantBadgeController mixin. 

How to test (on local dev):
1. [Create a custom badge ](http://localhost:4200/admin/badges/new)- this is the easiest way to get a grantable badge.
2. Navigate to a topic.
3. As an admin/mod, select the grant badge option from the wrench menu for a specific post.
4. Select the custom badge from the menu and click the Grant Badge button. A success flash should appear.
5. Select another post in the same topic and do steps 3-4. This tests that the selectedPost does change.

Testing Artifact:

https://github.com/discourse/discourse/assets/133760061/450d67a8-71ed-4e3e-8408-95955a9f6173
